### PR TITLE
[Bug Fix] GR1 & Gr3 - list the UPNs having non-enabled MFA

### PR DESCRIPTION
**Pull request by dutt0**
## Overview/Summary

This pull request fixes the reporting bug in 163gccspm for GR1 & Gr3 GA MFA UPN list feature. 

## This PR fixes/adds/changes/removes

In total, there should be 4 types of authentication methods to consider:
1.  #microsoft.graph.microsoftAuthenticatorAuthenticationMethod
2.  #microsoft.graph.phoneAuthenticationMethod    
3.  #microsoft.graph.passwordAuthenticationMethod 
4.  #microsoft.graph.emailAuthenticationMethod  

This PR has the fix for the MFA authentication by including two more authentication methods that users may have. Since dev tenant users did not have password and email authentication methods, this issue was not prominent in that tenant. However, the users in Test tenant carries these MFA. This pull request has the fix for this use case.

## Testing Evidence

image

## As part of this Pull Request I have

- [x] Checked for duplicate [Pull Requests](https://github.com/ssc-spc-ccoe-cei/azure-guardrails-solution-accelerator/pulls)
- [x] Associated it with relevant [GitHub Issues](https://github.com/ssc-spc-ccoe-cei/azure-guardrails-solution-accelerator/issues)
- [x] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/ssc-spc-ccoe-cei/azure-guardrails-solution-accelerator/tree/main)
- [x] Performed testing and provided evidence.
- [x] Updated relevant and associated documentation.
- [x] Ensure PowerShell module versions have been updated (manually or with the ./tools/Update-ModuleVersions.ps1 script)

### CODE REVIEW

**Review by alalvi00 — CHANGES_REQUESTED** [review]
The methods that you are checking for in your new code
 
#microsoft.graph.passwordAuthenticationMethod - This is users password and doesnt qualify as a second auth method for MFA
#microsoft.graph.emailAuthenticationMethod - This is users email used to reset user password and doesnt qualify as second auth method for MFA
 
 
The two methods above should not be checked for MFA as they don't represent the multi-factor that we are looking for hence why the old code didn't include them. We should check for authenticator app or phone authentication since theyre used as second auth method

If we use the new logic in the pr, then any user without authenticator app and phone number will be able to pass the compliancy test if they have a password and email used to reset

**Comment by dutt0** [comment]
@alalvi00 As you mentioned yesterday, you have #microsoft.graph.passwordAuthenticationMethod and #microsoft.graph.AuthenticatorAuthenticationMethod  and both are enabled for MFA. This logic goes by that perspective.

**Comment by alalvi00** [comment]
> @alalvi00 As you mentioned yesterday, you have #microsoft.graph.passwordAuthenticationMethod and #microsoft.graph.AuthenticatorAuthenticationMethod and both are enabled for MFA. This logic goes by that perspective.

Thats fine but a user will also pass if they only have password and email and I have tested this in the test tenant

**Comment by dutt0** [comment]
⁠Just for the reference in localexecution this is what I see for $commnts and the ComplianceStatus is also false. Thanks for your testing and giving reviews. I will discuss the logic in tomorrow's meeting and thereafter finalize the PR

Testcase #1 :
-<testuser>@xxxxxxxxx.onmicrosoft.com
-< ila's upn>@xxxxxxxxx.onmicrosoft.com

Testcase #2 :
-<testuser>@xxxxxxxxx.onmicrosoft.com
-<Ali's upn>@xxxxxxxxx.onmicrosoft.com

image

File: src/Guardrails-Common/GR-Common.psm1 [hunk 1]
```diff
@@ -396,30 +396,61 @@ function Check-GAAuthenticationMethods {
...
                                 MFAStatus   = $false
                                 MFAComments = $hiddenUPN
                                 
                             }
-                            #add the list to GA MFA list
+                            # add the list to GA MFA list
                             $GAUPNsMFA += $GAUPNtemplate
                         }
                     }
+                    # Condition: when user's authentication methods > 1
```
Comment (singhgss) [inline src/Guardrails-Common/GR-Common.psm1:428]: @dutt0  

We can make this code a little bit cleaner 

```
$targetAuthMethodTypes = @(
    "#microsoft.graph.phoneAuthenticationMethod",
    "#microsoft.graph.microsoftAuthenticatorAuthenticationMethod",
    "#microsoft.graph.passwordAuthenticationMethod",
    "#microsoft.graph.emailAuthenticationMethod"
)

$GAUPNtemplate = [PSCustomObject]@{
    UPN        = $globalAdminAccount
    MFAStatus  = $false
    MFAComments = $hiddenUPN
}

foreach ($authmeth in $authenticationmethods) {
    if ($targetAuthMethodTypes -contains $($authmeth.'@odata.type')) {
        $mfaCounter += 1
        break
    } else {
        # Clone the template and add to the list
        $GAUPNsMFA += $GAUPNtemplate.psobject.Copy()
    }
}
```



I think this introduces a bug what happens if first authMeth that comes in the iteration is none of the 4 auth methods, wouldn't this create a  GAUPNTemplate list and then in the next iteration it finds the one of the 4 auth methods and it also increases mfacounter 

So we end up posting message of mfa not enabled as well as showing message that it is enabled for that UPN

**Comment by dutt0** [comment]
Thanks for your reviews. I am refactoring this function. Cancelling this PR. I will raise another PR later

**Pull request closed without merging**